juju/application.py

```diff
@@ -684,6 +690,68 @@ async def refresh(
 
     upgrade_charm = refresh
 
+    async def local_refresh(
+            self, channel=None, force=False, force_series=False, force_units=False,
+            path=None, resources=None):
+        """Refresh the charm for this application with a local charm.
+
+        :param str channel: Channel to use when getting the charm from the
+            charm store, e.g. 'development'
+        :param bool force_series: Refresh even if series of deployed
+            application is not supported by the new charm
+        :param bool force_units: Refresh all units immediately, even if in
+            error state
+        :param str path: Refresh to a charm located at path
+        :param dict resources: Dictionary of resource name/filepath pairs
+        :param int revision: Explicit refresh revision
+        :param str switch: Crossgrade charm url
+
+        """
+        app_facade = self._facade()
+
+        charm_dir = os.path.abspath(
+            os.path.expanduser(path))
+        model_config = await self.get_config()
+
+        series = get_charm_series(charm_dir)
+        if not series:
+            model_config = await self.get_config()
+            default_series = model_config.get("default-series")
+            if default_series:
+                series = default_series.value
+
+        charm_url = await self.model.add_local_charm_dir(charm_dir, series)
+        if resources is not None:
+            if str(path).endswith('.charm'):
+                with zipfile.ZipFile(path, 'r') as charm_file:
+                    metadata = yaml.load(charm_file.read('metadata.yaml'), Loader=yaml.FullLoader)
+            else:
+                entity_path = Path(path)
+                metadata_path = entity_path / 'metadata.yaml'
+                metadata = yaml.load(metadata_path.read_text(), Loader=yaml.FullLoader)
```

Comment:
This feels like it could be extracted somewhere so we could reuse this.

Reply:
I agree, would this make sense somewhere like `utils.py`?

Reply:
I think we should be specific about this and put it in `juju/charm.py`

Reply:
Updated, I didn't add it as a class method as i don't think it makes sense to initialize a charm object in my use case,